websockets/py35/server.py

```diff
@@ -0,0 +1,21 @@
+async def __aenter__(self):
+    return await self
+
+
+async def __aexit__(self, exc_type, exc_value, traceback):
+    self.ws_server.close()
+    await self.ws_server.wait_closed()
+
+
+async def __await_impl__(self):
+    # Duplicated with __iter__ because __await__ must call coroutines with
+    # await in Python 3.7 while __iter__ must call them with yield from to
+    # preserve compatibility with Python 3.4.
+    server = await self._creating_server
+    self.ws_server.wrap(server)
+    return self.ws_server
+
+
+def __await__(self):
+    # I'm not finding a better way to take advantage of PEP 492.
```

Comment:
If you put the `async` in front of `__await__`, can you remove the `__await_impl__`?

Reply:
Nope. I tried :-)

I found this solution on StackOverflow, where someone dug it from a lib written by the author of PEP 492.

I find this very weird.

If you had time to ask on the asyncio mailing-list (where you seem to post occasionally) if this is the intended way to write an `__await__` method, that would be very useful.

As far as I understand (i.e. not much), `__await__` needs to return an type that is most easily obtained as the return value of calling `__await__` on an `async def` function.

Reply:
That type reports itself as `<class 'coroutine_wrapper'>` and I'm not sure what it is.